13_sandboxes/test_case_generator.py

```diff
@@ -0,0 +1,379 @@
+
+
+def get_sandbox_image(gh_owner: str, gh_repo_name: str):
+    ALLURE_VERSION = "2.34.1"
+    MODULE_URL = f"https://github.com/{gh_owner}/{gh_repo_name}"
+
+    image = (
+        modal.Image.debian_slim()
+        .apt_install("git", "curl", "tar", "default-jre")
+        .pip_install("webdiff")
+        .run_commands(
+            f"git clone {MODULE_URL}",
+            "curl -sSL https://install.python-poetry.org | python3 -",
+            "mkdir -p /opt/allure",
+            f"curl -sL https://github.com/allure-framework/allure2/releases/download/{ALLURE_VERSION}/allure-{ALLURE_VERSION}.tgz | tar xz -C /opt/allure --strip-components=1",
+        )
+        .env({"PATH": "$PATH:/root/.local/bin:/opt/allure/bin"})
+    )
+
+    return image
+
+
+def run_sandbox(image: modal.Image, file_name: str):
+    new_file_name = file_name.replace(".py", "_llm.py")
+
+    cmd = (
+        f"webdiff password-analyzer/tests/{file_name} /data/outputs/{file_name}  --host 0.0.0.0 --port 8001 &&"
+        + "cd password-analyzer && "
+        + "poetry install --no-root && "
+        + "poetry run pytest --alluredir allure-results || true && "
+        + f"cp /data/outputs/{file_name} tests/{new_file_name} && "
+        + "poetry run pytest --alluredir allure-results || true && "
+        + "allure serve allure-results --host 0.0.0.0 --port 8000"
+    )
+
+    sb = modal.Sandbox.create(
+        "sh",
+        "-c",
+        cmd,
+        app=app,
+        image=image,
+        volumes={
+            "/data": files_volume,
+        },
+        timeout=300,  # TODO: should we keep this as the default?
+        encrypted_ports=[8000, 8001],
+    )
+    return sb
+
+
+@app.local_entrypoint()
+def main(
+    gh_owner: str,
+    gh_repo_name: str,
+    gh_module_path: str,
+    gh_tests_path: str,
+    gh_branch: str,
+):
+    # Start server
+    sg_lang_server = TestCaseServer()
+
+    # Download files to volume
+    input_files = download_files_to_volume.remote(
+        folder_paths=[gh_module_path, gh_tests_path],
+        gh_owner=gh_owner,
+        gh_repo_name=gh_repo_name,
+        gh_branch=gh_branch,
+    )
+
+    # Initialize client and generate test files
+    generator = TestCaseClient(url=sg_lang_server.serve.get_web_url())
+    output_files = list(generator.generate.map(input_files))
+    print("Test case files generated successfully! Creating sandboxes...")
+
+    # Create sandboxes to run the generated test files
+    sandboxes = create_sandboxes(output_files, gh_owner, gh_repo_name)
+    poll_sandboxes(sandboxes)
+
+
+# # Addenda
+# The below functions are utility functions.
+def create_sandboxes(filenames: list[str], gh_owner: str, gh_repo_name: str):
+    file_to_sandbox: dict[str, modal.Sandbox] = {}
+    for filename in filenames:
+        print(f"Running sandbox for {filename}")
+        image = get_sandbox_image(gh_owner, gh_repo_name)
+        sb = run_sandbox(image, filename)
+        file_to_sandbox[filename] = sb
+    time.sleep(20)
```

Comment:
hack to make sure that the tunnels show up at the very end, after the noisy sandbox logs. don't think there's a sane way to silence those logs, but i feel like it's ok? kinda cool to see pytest running in the background in the sandboxes.